### PR TITLE
check if cache files are readable

### PR DESCRIPTION
**Pull request by butonic**
drop of water on https://github.com/owncloud/enterprise/issues/511, cc @blizzz @felixboehm @DeepDiver1975

### CODE REVIEW

**Comment by MorrisJobke** [comment]
:+1:

**Comment by butonic** [comment]
Should we maybe throw an exception if the cache is read only in getCacheDir()? https://github.com/owncloud/core/blob/master/lib/private/cache/fileglobal.php#L12

Maybe also cache the $cache_dir in a variable to do the string concatenation, is_dir() and then is_writeable() check only once per request?

File: lib/private/cache/file.php [hunk 1]
```diff
@@ -83,7 +83,7 @@ public function set($key, $value, $ttl=0) {
 
 	public function hasKey($key) {
 		$storage = $this->getStorage();
-		if ($storage && $storage->is_file($key)) {
+		if ($storage && $storage->isReadable($key)) {
```
Comment (DeepDiver1975) [inline lib/private/cache/file.php]: from my understanding we should keep the check on is_file

File: lib/private/cache/fileglobal.php [hunk 2]
```diff
@@ -52,7 +52,7 @@ public function set($key, $value, $ttl=0) {
 	public function hasKey($key) {
 		$key = $this->fixKey($key);
 		$cache_dir = self::getCacheDir();
-		if ($cache_dir && is_file($cache_dir.$key)) {
+		if ($cache_dir && is_readable($cache_dir.$key)) {
```
Comment (DeepDiver1975) [inline lib/private/cache/fileglobal.php]: same as above - keep the is_file check

**Comment by DeepDiver1975** [comment]
> Should we maybe throw an exception if the cache is read only in getCacheDir()? https://github.com/owncloud/core/blob/master/lib/private/cache/fileglobal.php#L12
> 
> Maybe also cache the $cache_dir in a variable to do the string concatenation, is_dir() and then is_writeable() check only once per request?

possible enhancements in the future

**Comment by DeepDiver1975** [comment]
@butonic this is no for 8.0 from my understanding - sorry

File: lib/private/cache/file.php (hunk 1 above)
Comment (butonic) [inline lib/private/cache/file.php]: please explain. using is_file lead to data loss, see https://github.com/owncloud/enterprise/issues/511

Comment (DeepDiver1975) [inline lib/private/cache/file.php]:
```php
if ($storage && $storage->isReadable($key) && $storage->is_file($key) ) {

**Comment by butonic** [comment]
@DeepDiver1975 I am ok with 8.0.1, problem can also be fixed by using memcached. This just prevents problems in the future.

**Comment by PVince81** [comment]
:+1:

**Comment by MorrisJobke** [comment]
:+1:

**Comment by MorrisJobke** [comment]
Backport in #14018

**Pull request merged**